### PR TITLE
Add example gRPC client

### PR DESCRIPTION
**Pull request by mansam**
This PR adds an example gRPC client that can be used with the provider running in gRPC server mode to run (a feature-limited subset of) analyzer rulesets. This is not a production tool. It is intended only to serve as a demonstration.

## Usage

The gRPC client will construct a provider config from its command line arguments and use that to initialize the provider.

The flags are:

* `--server`: address of provider server
* `--kubeconfig`: path to the kubeconfig with the coordinates to the cluster to be analyzed
* `--namespaces`: comma-delimited list of namespaces to analyze
* `--rules`: comma-delimited list of paths to rulesets
* `--stop`: indicates that the provider should be sent the signal to shut down after results are returned.

Example command invocations:

```sh
./bin/serve --port 8888
```

```sh
./bin/grpc --server 127.0.0.1:8888 --kubeconfig .kube/config --namespaces konveyor-tackle --rules rules/bestpractices
```

## Example output

```json
[
  {
    "name": "k8s Best Practices",
    "violations": {
      "lonely-pod": {
        "description": "Pod does not have an OwnerReference. Pod lifecycle should be managed by a resource such as a Deployment to ensure availability.",
        "category": "optional",
        "incidents": [
          {
            "uri": "https://api.crc.testing:6443/api/v1/namespaces/konveyor-tackle/pods/busybox",
            "message": "Pod does not have an OwnerReference. Pod lifecycle should be managed by a resource such as a Deployment to ensure availability.",
            "variables": {
              "apiVersion": "v1",
              "kind": "Pod",
              "name": "busybox",
              "namespace": "konveyor-tackle"
            }
          }
        ],
        "effort": 1
      },
      "low-replica-count": {
        "description": "Deployment has a low replica count",
        "category": "optional",
        "incidents": [
          {
            <snip>
```

### CODE REVIEW

File: examples/grpc/README.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    },
+    "proxyConfig": {
+    }
+  }]
+}
+```
+
+The name must be `k8s`, and the address should be the address on which the gRPC provider is listening. `kubeConfigPath` and `baseModulesPath` are paths that need to be accessible to the the server binary where ever it happens to be running.
+
+The `--rules` flag must be a path to a file containing a json object with an property named `ruleSets` containing an array of paths to analyzer rulesets compatible with the provider. Such a ruleset is included in this repository unde `rules/bestpractices`.
```
Comment (eemcmullan) [inline examples/grpc/README.md]: In analyzer, `--rules` is a string array for dirs containing rules and their rulesets (this is validated by the analyzer). For example `--rules=/dir1` `--rules=dir2` etc. This is optional, but I just wanted to point it out in the case of consistency across the providers.

File: examples/grpc/main.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	Description string               `yaml:"description,omitempty" json:"description,omitempty"`
+	Tags        []string             `yaml:"tags,omitempty" json:"tags,omitempty"`
+	Violations  map[string]Violation `yaml:"violations,omitempty" json:"violations,omitempty"`
+	Errors      map[string]string    `yaml:"errors,omitempty" json:"errors,omitempty"`
+	Unmatched   []string             `yaml:"unmatched,omitempty" json:"unmatched,omitempty"`
+	Skipped     []string             `yaml:"skipped,omitempty" json:"skipped,omitempty"`
+	Labels      []string             `yaml:"labels,omitempty" json:"labels,omitempty"`
+}
+
+type Violation struct {
```
Comment (eemcmullan) [inline examples/grpc/main.go:181]: Could also import from https://github.com/konveyor/analyzer-lsp/blob/main/output/v1/konveyor/violations.go#L97

Comment (mansam) [reply examples/grpc/main.go:181]: I tried to import the Ruleset and Violations from there originally, but marshaling the structures to json/yaml was causing problems. The yaml encoder was recursing until it blew up the stack, and the json encoder was failing due to a type it couldn't understand. Replacing them with these types solved that. Not sure if it's related to the MarshalJSON and MarshalYAML methods on the types in the analyzer library, or what.

File: examples/grpc/README.md (hunk 1 above)
Comment (mansam) [reply examples/grpc/README.md]: Good idea. This is just a client that's kind of a stand in for the analyzer for development purposes, but consistency here would be good.

**Pull request merged**